### DISCUSSION TITLE
Wazuh is not showing custom alerts in wazuh dashboard

### DISCUSSION THREAD

**Discussion by EToLLa** [wazuh/wazuh#15483] [Q&A] [answered]
Hello. I'm new with  wazuh and trying make custom rule to see log form app in wazuh dashboard.
I recieve my logs and see them in archive.log, but I can't get them to show up in dashboard.
Here log variant:
2022 Oct 31 10:52:46 (Vbank190) any->EventChannel {"win":{"system":{"providerName":"cpcsp","eventID":"100","level":"0","task":"0","keywords":"0x80000000000000","systemTime":"2022-10-31T07:52:46.445204900Z","eventRecordID":"40880","channel":"Application","computer":"test","severityValue":"UNKNOWN","message":"\"CryptoPro CSP. kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\\\.\\REGISTRY\\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1.\""},"eventdata":{"data":"kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\\\\\\\.\\\\REGISTRY\\\\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1."}}}

Here are my simple rules:

  <rule id="100043" level="3">
    <decoded_as>json</decoded_as>
    <field name="win.system.providerName">cpcsp</field>
    <description>CryptoPro CSP messages</description>
  </rule>
  <rule id="100045" level="3" overwrite="yes">
    <if_sid>100043</if_sid>
    <match>cpcsp</match>
    <description>cpcsp messages</description>
  </rule>

Ruleset test :
{"win":{"system":{"providerName":"cpcsp","eventID":"100","level":"0","task":"0","keywords":"0x80000000000000","systemTime":"2022-10-31T07:52:46.445204900Z","eventRecordID":"40880","channel":"Application","computer":"test","severityValue":"UNKNOWN","message":"\"CryptoPro CSP. kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\\\.\\REGISTRY\\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1.\""},"eventdata":{"data":"kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\\\\\\\.\\\\REGISTRY\\\\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1."}}}

**Phase 1: Completed pre-decoding.
	full event: '{"win":{"system":{"providerName":"cpcsp","eventID":"100","level":"0","task":"0","keywords":"0x80000000000000","systemTime":"2022-10-31T07:52:46.445204900Z","eventRecordID":"40880","channel":"Application","computer":"test","severityValue":"UNKNOWN","message":"\"CryptoPro CSP. kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\\\.\\REGISTRY\\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1.\""},"eventdata":{"data":"kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\\\\\\\.\\\\REGISTRY\\\\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1."}}}'

**Phase 2: Completed decoding.
	name: 'json'
	win.eventdata.data: 'kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\\\.\\REGISTRY\\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1.'
	win.system.channel: 'Application'
	win.system.computer: 'test'
	win.system.eventID: '100'
	win.system.eventRecordID: '40880'
	win.system.keywords: '0x80000000000000'
	win.system.level: '0'
	win.system.message: '"CryptoPro CSP. kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\.\REGISTRY\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1."'
	win.system.providerName: 'cpcsp'
	win.system.severityValue: 'UNKNOWN'
	win.system.systemTime: '2022-10-31T07:52:46.445204900Z'
	win.system.task: '0'

**Phase 3: Completed filtering (rules).
	id: '100045'
	level: '3'
	description: 'cpcsp messages'
	groups: '["CryptoPro"]'
	firedtimes: '1'
	mail: 'false'
**Alert to be generated.

I just need to see all logs from cpcsp in dashboard. Logs are coming from ossec agent via windows/applications log
Please help wih your advice.
Thanks in advance!

**Comment by pereyra-m** [comment]
Hello @EToLLa !

To help you better with this issue, could you please upload the custom rule you have created?

It seems that `wazuh-logtest` is giving you a correct output for your sample log but if you don't see the alert in the dashboard, we'd need first to check the `alerts.log` or `alerts.json` files. If you find the related alerts in these files, then we could have an issue with filebeat.

But if the alerts aren't present in the alerts files, then the logs you see in `archives.log` aren't triggering your custom rules.
Have you restarted your manager after creating them?

Regards.

**Comment by EToLLa** [comment]
Hello @pereyra-m 
Here is my rule file. I was trying to write different rules, but al of them don't work.
Logs, whch i need, are appearing in alerts.log and alerts.json.
Yes, i restarted wazuh-manager after changes, made in rules. I also restarted wazuh-indexer and wazuh-dashboard.
[local_rules.zip](https://github.com/wazuh/wazuh/files/10103113/local_rules.zip)

**Comment by pereyra-m** [comment]
Hi again @EToLLa !

Thank you for sharing your custom rules. However, I can't thoroughly test them because they rely on custom decoders. Would you upload them too?

Now, if you see these alerts in both `alerts.log` and `alerts.json` but not in the dashboard, are you applying the correct time filter? Try using a longer time frame

![time_frame](https://user-images.githubusercontent.com/65046601/204301342-aee11035-0e06-446b-87c1-ef178d3d3f06.png)


Also, we'll check the state of filebeat, because this component is in charge of sending the alerts to the indexer. Please share the output of these commands in the manager
- `systemctl status filebeat`
- `filebeat test output`

Regards.

**Comment by EToLLa** [comment]
Hello, @pereyra-m !
Thank's for your help!
Here is my decoder file. There is a lot of junk there, but in example higher i am using decoder "JSON", which is standart decoder.
My custom rules and decoders don't work either (
[local_decoder.zip](https://github.com/wazuh/wazuh/files/10111290/local_decoder.zip)

I'm using extended time, but still don't see alerts. (
![image](https://user-images.githubusercontent.com/95977686/204483058-5e4032e3-4a2a-463b-9806-458341adb48c.png)

Here is output of filebeat commands

● filebeat.service - Filebeat sends log files to Logstash or directly to Elasticsearch.
   Loaded: loaded (/lib/systemd/system/filebeat.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2022-11-14 18:53:19 MSK; 2 weeks 0 days ago
     Docs: https://www.elastic.co/products/beats/filebeat
 Main PID: 936 (filebeat)
    Tasks: 11 (limit: 4915)
   Memory: 37.6M
   CGroup: /system.slice/filebeat.service
           └─936 /usr/share/filebeat/bin/filebeat --environment systemd -c /etc/filebeat/filebeat.yml --path.home /usr/share/filebeat --path.config /etc/filebeat --path.data /var/lib/filebeat --

Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.
root@Monitor190:/home/wadmin# filebeat test output
elasticsearch: https://127.0.0.1:9200...
  parse url... OK
  connection...
    parse host... OK
    dns lookup... OK
    addresses: 127.0.0.1
    dial up... OK
  TLS...
    security: server's certificate chain verification is enabled
    handshake... OK
    TLS version: TLSv1.3
    dial up... OK
  talk to server... OK
  version: 7.10.2

**Comment by pereyra-m** [comment]
Hi @EToLLa !

Thank you all the information provided.
I don't see any issue in your filebeat service.

I've done the following tests on my environment with Wazuh v4.3.0 and everything is working as expected:
- I copied both files you shared (local rules and decoders).
- I've tested that they are working properly with `wazuh-logtest`
![alert_generated](https://user-images.githubusercontent.com/65046601/204685547-7d5f6d52-5a27-4248-b329-c0074f27c5d0.png)

- Added to `ossec.conf` in the manager a new block for reading a test file
```xml
  <localfile>
    <log_format>syslog</log_format>
    <location>/var/ossec/test.log</location>
  </localfile>
```
- Restart the manager `systemctl restart wazuh-manager`
- Add new lines to the test file
```bash
[root@wazuh-server ossec]# echo '2022 Oct 31 10:52:46 (Vbank190) any->EventChannel {"win":{"system":{"providerName":"cpcsp","eventID":"100","level":"0","task":"0","keywords":"0x80000000000000","systemTime":"2022-10-31T07:52:46.445204900Z","eventRecordID":"40880","channel":"Application","computer":"test","severityValue":"UNKNOWN","message":""CryptoPro CSP. kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\.\REGISTRY\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1.""},"eventdata":{"data":"kontur.plugin.host.exe:CPCAcquireContext (pszContainer=\\\\.\\REGISTRY\\0af3afa1-5c77-21ba-25e4-2b79787aac87, dwFlags=0x40). Result=1, Err=0x0. hProv=0x69b000e1."}}}' >> test.log 
```
- The alerts are available in the dashboard
![alert_in_dashboard](https://user-images.githubusercontent.com/65046601/204685825-c9cce8ec-6be5-489c-bfb0-919c68e9ddee.png)

This way we can confirm that the whole process works.
Regards.

**Comment by EToLLa** [comment]
Hi, @pereyra-m ! I'm glad, that my decoders and rules working. )
But my problem still there. I receive logs in my alerts.log  through windows ossec-agent , but they do not show in dashboard. When i tried to  make ny wazuh look in alerts.log ( i know it's stupid ), alerts started to show but this was vere hard for wazuh to look in to alerts.log a and it regulary crashed. What advice can you give ?

**Reply by pereyra-m** [reply]
Hello again @EToLLa !

Wazuh should publish in the dashboard all the alerts in `alerts.json/log` files. There is no need to configure these files to be monitored.

I'd like to know if you are able to see new alerts every day from the rest of the agents, is this the only type of event you are having trouble with? Does this happen with any other custom rule? Have you performed the same steps as me to test the full process?

Also, have you seen any error messages in the UI related to the indexes? Is your installation a single node or a cluster? What is the Wazuh version you are using? Does your host have enough available space?

Regards.

**Comment by EToLLa** [comment]
Hello? @pereyra-m ! Again, thank you for your help.
I made as you say and i received alert :
![image](https://user-images.githubusercontent.com/95977686/205261419-6161740c-38cf-4287-84b9-97931a642c29.png)
but the problem still remains. I see logs in archive.json , archive.log  , but do not receive them in alerts.json.
Yes, i see new alerts every day form different agents, almost all of them are default. Problem with some custom rules. For example my custom rule vmware working nice :
![image](https://user-images.githubusercontent.com/95977686/205262584-e473f3a9-0eb7-4e60-8853-c8d22d6bf137.png)
No i don't have any error messages related to the indexes. I have a single server with wazuh, not container. Wazuh version 4.3.3
My host has enough space ( 40gb free ).
The solution i see for now, is somehow put neccesery log from archive.log ( or archive.json ) in seperate log file.

**Comment by pereyra-m** [comment]
Hello @EToLLa !

Thank you very much again for all the information provided and the tests you've made.

I'll summarize some key points because after all these messages some basic details of Wazuh might not be clear.

- Wazuh receives information from the agents. If a log doesn't trigger an alerts, it is discarded.
- To save all logs, even if they don't trigger a rule, we have the [logall](https://documentation.wazuh.com/4.3/user-manual/reference/ossec-conf/global.html#logall) option. Enabling it saves everything in `/var/ossec/logs/archives/archives.log`. This is optional and might fill the disk really quickly.
- Having a log in `archives.log` but not seeing it in `alerts.log` means that it didn't match any rule, so you won't see it in the dashboard.
- Having a log in `archives.log` and seeing it in `alerts.log` means it generated an alert and we should see it in the dashboard soon. If we don't, we could have a problem in filebeat for example.

If we consider all this, when you say the following

> I made as you say and i received alert :

> but the problem still remains. I see logs in archive.json , archive.log , but do not receive them in alerts.json.

I only can think that the rule matches with the log when it's read from a file but not when it comes from directly from eventchannel. Specially if we don't have any error message.

We'll work then on the rules then.
First, even when it isn't directly linked to this issue, remove all the `overwrite="yes"` in your custom rules. This option is only required when you need to changue the behavior of a custom rule by using the same ID.

Also, in the `CryptoPro` group, there are some things that we could fix. I'll copy here the section 

```xml
<group name="CryptoPro">
  <rule id="100040" level="3">
    <decoded_as>CryptoPro</decoded_as>
    <match>Crypto</match>
    <description>CryptoPro rules</description>
  </rule>
  <rule id="100041" level="3" overwrite="yes">
     <if_sid>100040</if_sid>
     <description>Alert from CryptoPro</description>
  </rule>
  <rule id="100042" level="3" overwrite="yes">
    <if_sid>100040</if_sid>
    <decoded_as>CryptoPro-f</decoded_as>
    <description>CryptoPro alert</description>
  </rule>
  <rule id="100043" level="3">
    <decoded_as>json</decoded_as>
    <field name="win.system.providerName">cpcsp</field>
    <description>CryptoPro CSP messages</description>
  </rule>
  <rule id="100045" level="3" overwrite="yes">
    <if_sid>100043</if_sid>
    <match>cpcsp</match>
    <description>cpcsp messages</description>
  </rule>
  <rule id="100046" level="3" overwrite="yes">
    <match>Crypto</match>
    <description>Crypto messages</description>
  </rule>
</group>
```

- The rule **100041** has only one condition, the previous **100040** rule but it doesn't add a new condition . And it has the same level than it's parent rule. I'd remove it.
- The rule **100042** depends on **100040** but it adds another `decoded_as` field different than father rule. The child decoder `CryptoPro-f` only extracts information so I wouldn't use it in any `decoded_as` field.
- The rule **100046** has the same conditions than **100040**, I'd remove it.
- The rule **100045** depends on **100043** but it's parent already found the value `cpcsp` in `win.system.providerName`, so this rule might be redundant. 

I suggest you to read more in detail [Custom rules and decoders](https://documentation.wazuh.com/4.3/user-manual/ruleset/custom.html) and [Creating decoders and rules from scratch](https://wazuh.com/blog/creating-decoders-and-rules-from-scratch/).

If you need more help writing these rules, I'd need more example logs and a brief explanation of what you are trying to detect.
Regards.

**Comment by EToLLa** [comment]
Hello again, @pereyra-m !
there are many similar rules, because i was trying to make them working somehow. )
I need a simple rule at first - all logs with CryptoPro i need to see in dashboard.
I removed all decoders and all rules except rule 100043 which decoded as json. Restarted wazuh-manager. Tommorow i'll check, if it work or not.

**Reply by pereyra-m** [reply]
OK, I'll wait your update!
Regards.

**Comment by EToLLa** [comment, marked as answer]
If someone interested i solved my problem by making simple script, which copies logs i need from archive.log to a new file one time a day. Wazuh looks in this files, and my decoders and rules work. I know it isn't good solution, but it works for me.

#!/bin/bash

#Script to copy  logs from archive.log
cat /var/ossec/logs/archives/archives.log | while read y
do
grep *what u need* > /*path-to-new-file*.log
#echo $y > /*path-to-new-file*.log
done
